### ISSUE TITLE
v8-CPP-Inference can net.forward v5 not v8

### ISSUE THREAD

**Issue by Zomcxj** [ultralytics/ultralytics#8905]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

version ultralytics-8.1.0   torch 1.10.0  cuda11.3  onnx 1.12.0 
  
step1: export yolov8n.pt  opset 12 imgsz=640  
cannot cv::dnn::readNetFromONNX()

ERROR:0] global /home/daway/local/OpenCV452/opencv-4.5.2/modules/dnn/src/onnx/onnx_importer.cpp (2110) handleNode DNN/ONNX: ERROR during processing node with 2 inputs and 1 outputs: [Add]:(406)
terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.5.2) /home/daway/local/OpenCV452/opencv-4.5.2/modules/dnn/src/onnx/onnx_importer.cpp:2129: error: (-2:Unspecified error) in function 'handleNode'
> Node [Add]:(406) parse error: OpenCV(4.5.2) /home/daway/local/OpenCV452/opencv-4.5.2/modules/dnn/src/onnx/onnx_importer.cpp:916: error: (-215:Assertion failed) blob_0.size == blob_1.size in function 'handleNode'
> 
已放弃 (核心已转储)

step2 export yolov8n.pt  opset 10 imgsz=640  
can cv::dnn::readNetFromONNX()  
cannot net.forward(outputs, net.getUnconnectedOutLayersNames());

terminate called after throwing an instance of 'cv::Exception'
  what():  OpenCV(4.5.2) /home/daway/local/OpenCV452/opencv-4.5.2/modules/dnn/src/layers/../cuda4dnn/primitives/scale_shift.hpp:135: error: (-215:Assertion failed) 0 in function 'operator()'

i tried yolov5  opset 10 and 12   all can infer

**Comment by Zomcxj** [comment]
i tried opencv4.7.0  ok for yolov8n.onnx   one class   1,5,8400
but for yolov5s  when opencv4.5.2  is   1,25200,6     when  opencv4.7.0  is  1,3,80 
it's  up to opencv?   It cannot be inferred that v5 v8 is under the same opencv

**Comment by glenn-jocher** [comment]
@Zomcxj hello! It's great to hear that you've made progress with OpenCV 4.7.0 for YOLOv8n.onnx. Regarding the difference in output dimensions between YOLOv5 and YOLOv8 under different OpenCV versions, it indeed seems related to how OpenCV versions handle ONNX models. Different versions of OpenCV may interpret the ONNX graph differently, leading to variations in output dimensions.

For YOLOv5, the expected output dimension format is typically `[batch, num_classes + 5, num_anchors * grid_height * grid_width]`, which seems to align with what you observed in OpenCV 4.5.2. The change in dimensions with OpenCV 4.7.0 suggests a difference in how the model's output is parsed or a change in the model's ONNX export compatibility with newer OpenCV versions.

To ensure consistent behavior across different versions of OpenCV, you might consider explicitly checking the output tensor dimensions and adapting your post-processing code accordingly. This way, you can accommodate the differences and ensure your application works as expected regardless of the OpenCV version.

Here's a quick example of how you might adjust your code based on the output tensor dimensions:

```cpp
cv::Mat output = net.forward();
int size = output.total();

if (size == 25200 * 6) {  // Assuming YOLOv5 output format in OpenCV 4.5.2
    // Process output assuming [1, 25200, 6] dimensions
} else if (size == 3 * 80) {  // Assuming YOLOv5 output format in OpenCV 4.7.0
    // Process output assuming [1, 3, 80] dimensions
}
```

This approach requires a bit more code but increases robustness across different OpenCV versions. If you have further questions or run into more issues, feel free to ask!

**Comment by Zomcxj** [comment]
thank you!  i find the question in v5 for opencv4.7.0.     
When export v5,  i have four layer output 1,3,80,80,6 1,3,40,40,6 1,3,20,20,6 and 1,25200,6  
opencv4.7.0 will load all layers,  so i get the first layer 1,3,80,80,6    
i just add if(v5) get output[3].data  can solve the question  or  export only output  1,25200,6

But I still want to ask why opencv4.5.2 can't net.forward() v8

**Comment by glenn-jocher** [comment]
@Zomcxj hi there! 😊 It's great to hear you've found a solution for handling YOLOv5 outputs with OpenCV 4.7.0. Indeed, selecting the appropriate output layer, like `output[3].data` for YOLOv5, or exporting the model to produce only the desired output format can effectively solve the issue you encountered.

Regarding your question about OpenCV 4.5.2 not being able to `net.forward()` YOLOv8, it's likely due to compatibility issues between the ONNX model exported from YOLOv8 and the ONNX parser available in OpenCV 4.5.2. Newer versions of YOLO and ONNX may utilize features or layers that were not fully supported or correctly interpreted by the ONNX parser in OpenCV 4.5.2. Upgrading to OpenCV 4.7.0, as you've done, often resolves these issues because newer OpenCV versions include updates and fixes to the ONNX parser, improving compatibility with more recent ONNX models.

Here's a quick tip for selecting the last output layer in a more dynamic way, which might be helpful if the number of output layers varies:

```cpp
// Assuming 'net' is your loaded model
std::vector<cv::Mat> outputs;
net.forward(outputs, net.getUnconnectedOutLayersNames());
cv::Mat finalOutput = outputs.back(); // Gets the last output layer data
```

This approach automatically adapts to the number of output layers your model has, making your code more flexible. Keep experimenting and don't hesitate to reach out if you have more questions!
